src/communication/communicator.hpp

```diff
@@ -45,31 +48,68 @@ class communicator {
-        cell_gid_partition_(cell_gid_partition)
-    {}
-
-    cell_local_size_type num_groups_local() const
+    explicit communicator(
+            const recipe& rec,
+            const domain_decomposition& dom_dec,
+            const gid_prop_map& gid_props)
     {
-        return cell_gid_partition_.size();
-    }
-
-    void add_connection(connection con) {
-        EXPECTS(is_local_cell(con.destination().gid));
-        connections_.push_back(con);
-    }
+        using util::make_span;
+        num_domains_ = comms_.size();
+        num_local_groups_ = dom_dec.num_local_groups();
+
+        // Make a list of local gid.
+        std::vector<cell_gid_type> gids;
+        gids.reserve(dom_dec.num_local_cells());
+        for (auto i: make_span(0, num_local_groups_)) {
+            const auto& group = dom_dec.get_group(i);
+            for (auto gid: group.gids()) {
+                gids.push_back(gid);
+            }
+        }
 
-    /// returns true if the cell with gid is on the domain of the caller
-    bool is_local_cell(cell_gid_type gid) const {
-        return algorithms::in_interval(gid, cell_gid_partition_.bounds());
-    }
+        // Count the number of local connections (i.e. connections terminating on this domain)
+        //  -> n_cons: scalar
+        // Calculate and store domain id of the presynaptic cell on each local connection
+        //  -> ps_doms: array with one entry for every local connection
+        // Also the count of presynaptic sources from each domain
+        //  -> ps_cnts: array with one entry for each domain
+        cell_local_size_type n_cons = 0;
+        std::vector<unsigned> ps_doms;
+        std::vector<cell_size_type> ps_cnts(num_domains_);
+        for (auto gid: gids) {
+            const auto conns = rec.connections_on(gid);
+            n_cons += conns.size();
+            for (auto con: conns) {
+                const auto src = dom_dec.gid_domain(con.source.gid);
+                ps_doms.push_back(src);
+                ps_cnts[src]++;
+            }
+        }
 
-    /// builds the optimized data structure
-    /// must be called after all connections have been added
-    void construct() {
-        if (!std::is_sorted(connections_.begin(), connections_.end())) {
-            threading::sort(connections_);
+        // Construct the connections.
+        // The loop above gave the information required to construct in place
+        // the connections as partitioned by the domain of their source gid.
+        connections_.resize(n_cons);
+        connection_part_ = algorithms::make_index(ps_cnts);
+        auto offsets = connection_part_;
+        std::size_t pos = 0;
+        for (auto gid: gids) {
+            const auto lg = gid_props.get(gid)->local_group;
+            for (auto c: rec.connections_on(gid)) {
+                const auto i = offsets[ps_doms[pos]]++;
+                connections_[i] = {c.source, c.dest, c.weight, c.delay, lg};
+                ++pos;
+            }
```

Comment:
Rather than have both a range-based `gid` loop and a separate `pos` counter, it would be clearer just to have an indexed loop with `pos`, or store all the gid information in the one vector and iterate over that.